### PR TITLE
pod restart when init containers are changed

### PR DESCRIPTION
**Pull request by sofat1989**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
A user updates the Pod specification, causing the init container image to change. Any changes to the init container image restarts the Pod. App container image changes only restart the app container.

**Which issue(s) this PR fixes**:

Fixes #
https://github.com/kubernetes/kubernetes/issues/82523
**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Review by mattjmcnaughton — COMMENTED** [review]
Thanks for your pr :)

I will give this a full look tomorrow, but as a jumping off point, could I ask you to add unit test coverage so we can be sure that this will continue to work going forward?

**Comment by krmayankk** [comment]
@kubernetes/sig-apps-pr-reviews  for this review as well
/ok-to-test

File: pkg/kubelet/kuberuntime/kuberuntime_container.go [hunk 1]
```diff
@@ -703,17 +703,24 @@ func (m *kubeGenericRuntimeManager) purgeInitContainers(pod *v1.Pod, podStatus *
...
 	for i := len(pod.Spec.InitContainers) - 1; i >= 0; i-- {
 		container := &pod.Spec.InitContainers[i]
 		status := podStatus.FindContainerStatusByName(container.Name)
-		if status != nil && isInitContainerFailed(status) {
-			return status, container, false
+		if status != nil {
+			expectedHash, actualHash, changed := containerChanged(container, status)
+			if changed {
+				klog.V(4).Infof("Init container spec hash changed (%d vs %d).", actualHash, expectedHash)
+				return nil, nil, false, true
```
Comment (krmayankk) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: should this logic live inside findNextInitContainerToRun  ?

**Comment by krmayankk** [comment]
Agree we need a UT. Before that I am not sure this is the desired behavior . CC'ing some people to get their review.

/assign @Random-Liu  @tallclair

File: pkg/kubelet/kuberuntime/kuberuntime_container.go (hunk 1 above)
Comment (sofat1989) [reply pkg/kubelet/kuberuntime/kuberuntime_container.go]: leaving this logic here is to re-use the `for` loop inside findNextInitContainerToRun. I suppose that `findNextInitContainerToRun` is to compute next actions for init containers. So i think it is fine that we have hash checking here.

**Comment by sofat1989** [comment]
@mattjmcnaughton @krmayankk Thanks for reviewing. I add one unit test. When the restart policy of the pod is `Always`, if the init container's image is changed, the we will get `killPod` which is true.

**Comment by krmayankk** [comment]
@sofat1989  are you still working on this ?

**Comment by sofat1989** [comment]
@krmayankk  it is pending for reviewing and feedback.  I want to make sure that this change is expected in the design.

**Pull request closed without merging**